src/torch_models.py

```diff
@@ -1,153 +1,323 @@
-"""Models useful for classification/regression."""
+"""Models useful for classification/regression.
+
+Note: to promote modularity, this file should NOT import CFG.
+"""
 
 import abc
 import logging
 import os
 import tempfile
 from dataclasses import dataclass
-from typing import List, Optional, Sequence, Tuple
+from typing import Callable, List, Sequence, Tuple
 
 import numpy as np
 import torch
 import torch.nn.functional as F
 from torch import Tensor, nn, optim
 
-from predicators.src.settings import CFG
 from predicators.src.structs import Array, Object, State
 
 torch.use_deterministic_algorithms(mode=True)  # type: ignore
 
+################################ Base Classes #################################
+
 
 class Regressor(abc.ABC):
-    """ABC for regressor classes."""
+    """ABC for regressor classes.
+
+    All regressors normalize the input and output data.
+    """
+
+    def __init__(self, seed: int) -> None:
+        self._seed = seed
+        self._rng = np.random.default_rng(self._seed)
+        # Set in fit().
+        self._x_dim = 0
+        self._y_dim = 0
+        self._input_shift = np.zeros(1, dtype=np.float32)
+        self._input_scale = np.zeros(1, dtype=np.float32)
+        self._output_shift = np.zeros(1, dtype=np.float32)
+        self._output_scale = np.zeros(1, dtype=np.float32)
 
-    @abc.abstractmethod
     def fit(self, X: Array, Y: Array) -> None:
-        """Train regressor on the given data.
+        """Train the regressor on the given data.
 
         X and Y are both two-dimensional.
         """
-        raise NotImplementedError("Override me")
+        num_data, self._x_dim = X.shape
+        _, self._y_dim = Y.shape
+        assert Y.shape[0] == num_data
+        logging.info(f"Training {self.__class__.__name__} on {num_data} "
+                     "datapoints")
+        X, self._input_shift, self._input_scale = _normalize_data(X)
+        Y, self._output_shift, self._output_scale = _normalize_data(Y)
+        self._fit(X, Y)
 
-    @abc.abstractmethod
-    def predict(self, arr: Array) -> Array:
+    def predict(self, x: Array) -> Array:
         """Return a prediction for the given datapoint.
 
-        arr is single-dimensional.
+        x is single-dimensional.
         """
-        raise NotImplementedError("Override me")
+        assert x.shape == (self._x_dim, )
```

Comment:
even before this, assert fit was called by checking that `self._x_dim > 0 and self._y_dim > 0`? (similar for the classifier below)